ELEC-257 LSC: Remove cyclic dependencies in libraries

ckitagawa opened the pull request:
This was achieved by moving the enums for `interrupt.h` and the `status` module to a new library called libcore which encompasses code shared between x86, STDPeriphLib and ms-common libraries. I have run all the tests on x86 and will ensure all stm32f0xx tests pass when I am working in the bay tomorrow before merging. 

The alternative solution if there are concerns about the scale/impact of this change is to: 
*   Remove the dependency on `status` from the files in question and use integers instead. 
*   Manually redefine the enums for interrupt priority, type and edge in their respective architecture interrupt files. (I would rather not do this as it requires more guard clauses but I would be willing to do it if it was significantly preferred).

Both would be acceptable to me although I find this solution to be simpler and allows the use of `status` in other libraries which I find appealing.

tituschow commented:
I think a shared core library makes sense to me. I think data structures that don't rely on anything else (FSM, FIFOs, etc.) could probably belong there too. We could just use a config header to define critical sections.

tituschow commented on projects/driver_controls/rules.mk:
```diff
@@ -6,7 +6,7 @@
 
 # Specify the device library you want to include
-$(T)_DEPS := ms-common
+$(T)_DEPS := ms-common libcore
```
If ms-common depends on libcore, you don't need to define it here.

ckitagawa replied:
Done.

ckitagawa commented:
When I do the library cleanup/documentation PRs I will consider moving some of the data structures over as well in that case :)

tituschow commented on libraries/ms-common/inc/gpio_it.h:
```diff
@@ -4,7 +4,7 @@
 
 #include "gpio.h"
-#include "interrupt.h"
+#include "interrupt_def.h"
```
I don't think we need to explicitly include interrupt_def.h? It should be included with stm32f0xx_interrupt.h

ckitagawa replied:
It's necessary for the interface definition since this is just the architecture universal header we don't include the architecture specific interrupt headers.
`StatusCode gpio_it_register_interrupt(GPIOAddress *address, InterruptSettings *settings,
                                      InterruptEdge edge, gpio_it_callback callback, void *context);`

tituschow replied:
Okay, sounds fine to me.

tituschow commented:
Saw this in your build log for stm32f0xx: Might want to double check and make sure this isn't breaking anything.
```
/usr/bin/../lib/gcc/arm-none-eabi/6.3.1/../../../../arm-none-eabi/bin/ld: warning: cannot find entry symbol Reset_Handler; defaulting to 0000000008000000
```

ckitagawa commented:
Noted, I will investigate tomorrow since I left my STM hardware in the bay.

ckitagawa commented:
Updates:

*   Moved tests in libcore to ms-common as stm32f0xx needs to be built prior to any tests as the `stm32f0xx_startup.s` file needs to be linked first. This would result in a circular dependency once again if the tests did not remain in ms-common or another library that is built after the `$(PLATFORM_LIB)` is included.
*   Fixed `test_soft_timers.c` which had a bug with a missing critical section.
*   Fixed `test_gpio.c` which was testing an undefined hardware behavior.
*   Moved `misc.h` to libcore.
